planner: refactor join-recorder

undertaker86001 opened the pull request:
Issue Number: close #62866 

Problem Summary:

### What changed and how does it work?

### Check List

Tests <!-- At least one of them must be included. -->

- [x] Unit test
- [x] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No need to test
  > - [ ] I checked and no code files have been changed.
  > <!-- Or your custom  "No need to test" reasons -->

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

<!-- compatibility change, improvement, bugfix, and new feature need a release note -->

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

## Core Significance of the Join Reorder Rule

Join Reorder is a key rule in database query optimization. Its goal is **to find the optimal table join order to minimize query execution cost**.

### Why Do We Need Join Reorder?

Imagine you need to query three tables: A, B, and C:
```sql
SELECT * FROM A JOIN B ON A.id = B.a_id JOIN C ON B.id = C.b_id
```

Different join orders can result in completely different execution costs:
- Order 1: (A JOIN B) JOIN C
- Order 2: A JOIN (B JOIN C)
- Order 3: (A JOIN C) JOIN B (if join conditions allow)

## Refactoring Plan Architecture

### 1. Rule Interface Layer

After refactoring, `JoinReorderRule` implements the standard `LogicalOptRule` interface, allowing it to be invoked by a unified optimization framework. 

### 2. Join Group Extraction

The system first needs to identify which tables can be reordered. 

The core logic is in the `extractJoinGroup` function, which:
- Recursively traverses the query plan tree
- Identifies consecutive inner joins
- Extracts equi-join and other filter conditions
- Checks for the presence of outer joins (which restrict reorder possibilities)

### 3. Two Solution Algorithms

#### Dynamic Programming Algorithm (DP Algorithm) 
Suitable for small-scale joins (≤6 tables):
- Uses bitmask to represent table subsets
- Finds the global optimum via dynamic programming
- Time complexity: O(3^n), where n is the number of tables

Core idea: `bestPlan[S] = min(Join(bestPlan[S1], bestPlan[S2]))` where S1 ∪ S2 = S

#### Greedy Algorithm [6](#5-5)

Suitable for large-scale joins (>6 tables):
- Starts from the table with the lowest cost
- Each time, selects the next table that yields the minimal cumulative cost
- Time complexity: O(n²)

### 4. Cost Calculation Model

The system uses a simplified cost model:

```
Cumulative Join Cost = CumCount(left table) + CumCount(right table) + RowCount(join result)
```

This formula considers:
- Processing cost of left/right subtrees
- The cost of the join operation itself (based on the number of result rows)

## Technical Advantages of Refactoring

### 1. Clear Dependency Relationships

Before refactoring, all logic was mixed in the `planner/core` package. After refactoring:
- `core/rule` package: rule interface implementation
- `core/rule/util` package: reusable utility functions
- `core/logicalop` package: logical operator definitions

This avoids circular dependencies and follows good software architecture principles.

### 2. Improved Code Reusability

`BaseSingleGroupJoinOrderSolver` is extracted as a base class. Both DP and greedy algorithms can reuse:
- Join condition checking logic
- Join construction logic
- Cost calculation logic

### 3. Enhanced Extensibility

Through the `JoinOrderSolver` interface, new join reorder algorithms (such as machine-learning-based optimizers) can be easily added in the future.

## Actual Execution Process

1. **Rule Matching**: Check if the query plan contains reorderable joins
2. **Group Extraction**: Identify consecutive inner join groups
3. **Algorithm Selection**: Choose DP or greedy algorithm based on the number of tables
4. **Cost Optimization**: Find the optimal join order
5. **Plan Reconstruction**: Replace the original plan with the optimized join order [9](#5-8)

undertaker86001 committed:
refactor join-recorder

Commit:
refactor join-recorder

undertaker86001 committed:
Merge remote-tracking branch 'origin/master'

undertaker86001 committed:
solve test cycle deps

undertaker86001 committed:
solve test cycle deps

undertaker86001 committed:
impl replaceJoinGroup

undertaker86001 committed:
solve cycle deps

undertaker86001 committed:
solve cycle deps

lance6716 commented:
/ok-to-test

Hi, please follows our PR template https://github.com/pingcap/tidb/blob/master/.github/pull_request_template.md
And sign CLA as bot comment

undertaker86001 committed:
add go test

undertaker86001 commented:
/check-cla

undertaker86001 committed:
delete useless deps

undertaker86001 committed:
remove useless deps

undertaker86001 committed:
solve cycle deps

undertaker86001 committed:
refactor join_recorder

undertaker86001 committed:
add untraker type

undertaker86001 committed:
impl todo function

### 1. **generateLeadingJoinGroup** - Complete Leading Hint Processing
- Fully implemented leading hint parsing logic
- Supports table alias matching
- Handles query block alias processing
- Processes restrictions for outer joins
- Constructs the leading join group

### 2. **Tracing Related Functions** - Complete Tracing Functionality
- `traceJoinReorder`: Traces join reorder operations
- `appendLogicalJoinCost`: Records logical join costs
- `appendJoinReorderTraceStep`: Adds tracing steps
- `allJoinOrderToString`: Converts plan trace to string
- `joinOrderToString`: Converts join order to string representation
- `extractJoinAndDataSource`: Extracts join and data source operators
- `simplify`: Simplifies plan trace, retaining only joins and data sources
- `findRoots`: Finds root nodes

### 3. **DP and Greedy Algorithms** - Complete Algorithm Implementation
- Fully implemented dynamic programming algorithm (`joinReorderDPSolver`)
- Fully implemented greedy algorithm (`joinReorderGreedySolver`)
- Includes graph construction, BFS, DP state transitions, and complete logic
- Supports connection checks, cost calculation, and other core functions

### 4. **Helper Functions** - Complete Utility Functions
- `findNodeIndexInGroup`: Finds node index in a group
- `nodesAreConnected`: Checks if nodes are connected
- `newJoinWithEdge`: Creates a join with an edge
- `makeBushyJoin`: Creates a bushy join
- `bfsGraph`: BFS graph traversal

### 5. **Cost Calculation and Statistics** - Complete Cost Model
- `baseNodeCumCost`: Basic node cumulative cost calculation
- `calcJoinCumCost`: Join node cumulative cost calculation
- Supports recursive statistics for child node costs

### 6. **Connection Checking and Handling** - Complete Connection Logic
- `checkConnection`: Checks if two plans can be connected
- `canApplyEqualCondition`: Checks if equality conditions can be applied
- `columnsInSchema`: Checks if columns are in the schema
- `makeJoin`: Creates join operations

### 7. **Join Creation Factory** - Complete Join Construction
- `createJoin`: Creates a new LogicalJoin
- `newCartesianJoin`: Creates a Cartesian join
- Supports various join types and conditions

---

undertaker86001 committed:
remove useless variable

AilinKid commented:
<img width="1226" height="420" alt="image" src="https://github.com/user-attachments/assets/f67a5a0b-1826-4049-b2fa-78d289063118" />
there are some ref error inside, you can tell from the build link, do you mind fix it?

undertaker86001 commented:
> <img alt="image" width="1226" height="420" src="https://private-user-images.githubusercontent.com/15825830/477886924-f67a5a0b-1826-4049-b2fa-78d289063118.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTU1ODI5ODgsIm5iZiI6MTc1NTU4MjY4OCwicGF0aCI6Ii8xNTgyNTgzMC80Nzc4ODY5MjQtZjY3YTVhMGItMTgyNi00MDQ5LWIyZmEtNzhkMjg5MDYzMTE4LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA4MTklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwODE5VDA1NTEyOFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTE0N2M3ZjkyZGUxYWY4YjZiNjQ0MDJkZTNiMmJkMzkzMjBlYmVjMzE2ZTEwMzk5NzU0MjRlOGY5NjY5MzNhNjkmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.4BJ3v-upMbAznvyuhcaARP_a8ctMr3d3OtrfLgGcQ3Q"> there are some ref error inside, you can tell from the build link, do you mind fix it?

I've been busy lately, I'll try to start repairing it over the weekend
